### PR TITLE
blockfilter: Avoid out-of-bounds script access.

### PR DESCRIPTION
**Pull request by jimpo**
Caught during review of #12254 by @TheBlueMatt. https://github.com/bitcoin/bitcoin/pull/12254#discussion_r212830981

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

Very nice find @TheBlueMatt

**Comment by laanwj** [comment]
utACK 40a9c6c38c6aeaac83fe2639c9ac2d2b262296fc

File: src/blockfilter.cpp [hunk 1]
```diff
@@ -208,7 +208,7 @@ static GCSFilter::ElementSet BasicFilterElements(const CBlock& block,
     for (const CTransactionRef& tx : block.vtx) {
         for (const CTxOut& txout : tx->vout) {
             const CScript& script = txout.scriptPubKey;
-            if (script[0] == OP_RETURN) continue;
+            if (script.size() > 0 && script[0] == OP_RETURN) continue;
             elements.emplace(script.begin(), script.end());
```
Comment (gmaxwell) [inline src/blockfilter.cpp]: Now it looks like it adds size 0 entries to the filter, which is a violation of BIP158.

I believe this should be "script.size() < 1 ||". No?

Comment (promag) [reply src/blockfilter.cpp]: nit to @gmaxwell comment, `script.empty() || ...`.

File: src/test/blockfilter_tests.cpp [hunk 2]
```diff
@@ -67,6 +67,7 @@ BOOST_AUTO_TEST_CASE(blockfilter_basic_test)
     CMutableTransaction tx_2;
     tx_2.vout.emplace_back(300, included_scripts[2]);
     tx_2.vout.emplace_back(0, excluded_scripts[0]);
+    tx_2.vout.emplace_back(600, included_scripts[5]); // Script is empty
```
Comment (promag) [inline src/test/blockfilter_tests.cpp]: So this should be `excluded_scripts[2]`?

nit, above in L62 add something along
```cpp
// This script is left empty.
// excluded_scripts[2];
```

**Comment by promag** [comment]
utACK f055995.

**Comment by gmaxwell** [comment]
utACK

**Comment by laanwj** [comment]
re-utACK f05599557a8305d16bd5965921583af9d012fc27

**Pull request merged**